docs/proposals/high-availability.md

```diff
@@ -0,0 +1,34 @@
+# High Availability of Daemons in Kubernetes
+This document serves as a proposal for high availability of the master daemons in kubernetes.
+
+## Design Options
+1. Hot Standby Daemons: In this scenario, data and state are shared between the two deamons such that an immediate failure in one daemon causes the the standby deamon to take over exactly where the failed daemon had left off.  This would be an ideal solution for kubernetes, however it poses a series of challenges in the case of controllers where daemon-state is cached locally and not persisted in a transactional way to a storage facility.  As a result, we are **NOT** planning on this approach. 
+
+2. **Cold Standby Daemons**: In this scenario there is only one active daemon acting as the master and additional daemons in a standby mode.  Data and state are not shared between the active and standby daemons, so when a failure occurs the standby daemon that becomes the master must determine the current state of the system before resuming functionality.
+
+3. Stateless load-balanced Daemons: Stateless daemons, such as the apiserver, can simply load-balance across any number of servers that are currently running.  Their general availability can be continuously updated, or published, such that load balancing only occurs across active participants.  This aspect of HA is outside of the scope of *this* proposal because there is already a partial implementation in the apiserver.
```

Comment:
> On Apr 17, 2015, at 3:55 PM, David Oppenheimer <notifications@github.com> wrote:
> 
> In docs/proposals/high-availability.md:
> 
> > @@ -0,0 +1,34 @@
> > +# High Availability of Daemons in Kubernetes
> > +This document serves as a proposal for high availability of the master daemons in kubernetes.
> > +
> > +## Design Options
> > +1. Hot Standby Daemons: In this scenario, data and state are shared between the two deamons such that an immediate failure in one daemon causes the the standby deamon to take over exactly where the failed daemon had left off.  This would be an ideal solution for kubernetes, however it poses a series of challenges in the case of controllers where daemon-state is cached locally and not persisted in a transactional way to a storage facility.  As a result, we are **NOT** planning on this approach. 
> > +
> > +2. **Cold Standby Daemons**: In this scenario there is only one active daemon acting as the master and additional daemons in a standby mode.  Data and state are not shared between the active and standby daemons, so when a failure occurs the standby daemon that becomes the master must determine the current state of the system before resuming functionality.
> > +
> > +3. Stateless load-balanced Daemons: Stateless daemons, such as the apiserver, can simply load-balance across any number of servers that are currently running.  Their general availability can be continuously updated, or published, such that load balancing only occurs across active participants.  This aspect of HA is outside of the scope of *this* proposal because there is already a partial implementation in the apiserver.
> Is it true that a client can send a request to any API server and things will just work? I thought that the plan was to do sharding (a somewhat different model than any-request-can-go-anywhere) and that it wasn't supported yet.
> 
> BTW I'm wondering if the property is really "stateless"ness since the API server does write to etcd (unless the argument for calling it stateless is that it is just a conduit between etcd and clients, and doesn't have any of its own logic?).
> 
It's not uncommon to say the apiserver is a stateless "tier", because etcd holds the state.
> —
> Reply to this email directly or view it on GitHub.
>